### PR TITLE
Upgrade to Mongoid 5.2

### PR DESCRIPTION
**Pull request by jakolehm**
This upgrade will bring Mongo >= 3.2 compatibility (because moped will change to official mongodb driver).

https://docs.mongodb.com/ruby-driver/master/reference/driver-compatibility/

### CODE REVIEW

**Comment by jakolehm** [comment]
@SpComb what should we do with `MopedSessionTracer`?

**Comment by SpComb** [comment]
> @SpComb what should we do with MopedSessionTracer?

Need to look at the `Mongo` driver and how it handles thread-safety first. The [`MopedSessionTracer`](https://github.com/kontena/kontena/blob/69aff1dd6aab1c3759b9f97fc48e83709739351d/server/lib/moped_session_tracer.rb) is very simple, it just hooks `initialize` and any method that later gets called for each query.

**Comment by jakolehm** [comment]
Rebased.

**Comment by jnummelin** [comment]
@SpComb @daniel-bytes Please take look, we want to get this in early for 1.3.0 to get it properly battle-tested :)

**Comment by jakolehm** [comment]
> Is there some breaking things in 6 series, can't we jump on that?

Enum does not support 6 series.

File: server/Gemfile [hunk 1]
```diff
@@ -5,9 +5,9 @@ gem 'roda', '~> 2.5.0'
 gem 'celluloid', '~> 0.17.3'
 gem 'tilt-jbuilder'
 gem 'mutations'
-gem 'mongoid', '~> 4.0.0'
+gem 'mongoid', '~> 5.1'
```
Comment (SpComb) [inline server/Gemfile]: Perhaps `~> 5.2`, or `~> 5.2.1` to lock it to the `5.2.x` releases?

**Comment by SpComb** [comment]
Seems to be some 
```
I, [2017-05-11T16:16:47.196860 #7190]  INFO -- WebsocketBackend: node CAR3:26EI:CKHW:M3US:6L3D:W7CI:Y6KU:CJ5D:6ERB:DUHB:2ZAB:6Y3Q connection closed
I, [2017-05-11T16:16:47.198614 #7190]  INFO -- Agent::NodeUnplugger: disconnect node core-02
W, [2017-05-11T16:16:47.218775 #7190]  WARN -- : MONGODB | Unsupported client option 'safe'. It will be ignored.
W, [2017-05-11T16:16:47.220049 #7190]  WARN -- : MONGODB | Unsupported client option 'safe'. It will be ignored.
W, [2017-05-11T16:16:47.221809 #7190]  WARN -- : MONGODB | Unsupported client option 'safe'. It will be ignored.
W, [2017-05-11T16:16:47.232352 #7190]  WARN -- : MONGODB | Unsupported client option 'safe'. It will be ignored.
W, [2017-05-11T16:16:47.236416 #7190]  WARN -- : MONGODB | Unsupported client option 'safe'. It will be ignored.
W, [2017-05-11T16:16:47.237936 #7190]  WARN -- : MONGODB | Unsupported client option 'safe'. It will be ignored.
```

Perhaps from this?
```
server/app/services/agent/node_unplugger.rb:        c.with(safe: false).set(:deleted_at => deleted_at)
```

**Comment by SpComb** [comment]
There seems to be something going on with labels in the `ContainerStat` collection?

```
E, [2017-05-11T16:17:53.543344 #7198] ERROR -- RpcServer: BSON::String::IllegalKey: 'io.kontena.container.deploy_rev' is an illegal key in MongoDB. Keys may not start with '$' or contain a '.'.
E, [2017-05-11T16:17:53.543422 #7198] ERROR -- RpcServer: /home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/string.rb:69:in `to_bson_key'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:45:in `block in to_bson'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `each'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `to_bson'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:46:in `block in to_bson'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `each'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `to_bson'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:46:in `block in to_bson'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `each'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `to_bson'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:49:in `block in to_bson'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:46:in `each'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:46:in `each_with_index'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/array.rb:46:in `to_bson'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:46:in `block in to_bson'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `each'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/bson-4.2.1/lib/bson/hash.rb:43:in `to_bson'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/serializers.rb:163:in `serialize'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/message.rb:213:in `block in serialize_fields'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/message.rb:201:in `each'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/message.rb:201:in `serialize_fields'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/protocol/message.rb:104:in `serialize'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:193:in `block in write'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:192:in `each'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:192:in `write'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:162:in `deliver'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:108:in `block in dispatch'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/monitoring/publishable.rb:47:in `publish_command'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection.rb:107:in `dispatch'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/write/command/writable.rb:39:in `block in execute'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server/connection_pool.rb:107:in `with_connection'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/server.rb:242:in `with_connection'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/write/command/writable.rb:38:in `execute'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/write/bulk/insert.rb:59:in `execute_write_command'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/operation/write/bulk/bulkable.rb:39:in `execute'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/bulk_write.rb:182:in `insert_one'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/bulk_write.rb:153:in `execute_operation'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/bulk_write.rb:61:in `block (2 levels) in execute'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/bulk_write.rb:60:in `each'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/bulk_write.rb:60:in `block in execute'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/retryable.rb:112:in `write_with_retry'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/bulk_write.rb:56:in `execute'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/collection.rb:404:in `bulk_write'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongo-2.4.1/lib/mongo/collection.rb:382:in `insert_many'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/mongoid-5.2.1/lib/mongoid/query_cache.rb:182:in `insert_many_with_clear_cache'
/home/kontena/kontena/kontena/server/app/services/rpc/container_handler.rb:134:in `flush_stats'
/home/kontena/kontena/kontena/server/app/services/rpc/container_handler.rb:109:in `stat'
/home/kontena/kontena/kontena/server/app/services/rpc_server.rb:85:in `handle_notification'
/home/kontena/kontena/kontena/server/app/services/rpc_server.rb:43:in `process!'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
/home/kontena/kontena/kontena/server/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
```

**Comment by SpComb** [comment]
The `GridServiceDeploy#abort!` merged in from git master (not present in this PR branch!) calls `self.set(:deploy_state => :error)` without the leading `_`... and the specs for `expect(subject).to be_error` still pass. Change the spec to `expect(subject.reload).to be_error`, and now it fails :/

The breaking mongoid-enum changes for `set(...)`  are painful... the specs are do not catch them, because it seems to update the enum value on the document object, but it does not update it in mongo :face_with_head_bandage:

**Comment by SpComb** [comment]
>> Is there some breaking things in 6 series, can't we jump on that?
> 
> Enum does not support 6 series.

We could also consider dropping `Mongoid::Enum`, it's only used in four models (`GridDomainAuthorization`, `GridServiceDeploy`, `GridServiceInstanceDeploy`, `StackDeploy`), and the upstream project seems to be a bit dead (although there are some updated forks).

OTOH it would need some migration to rename the `_state` fields back to just `state` to keep the code sane...

**Comment by SpComb** [comment]
The `Mongoid::Enum` `set(:_deploy_state => ...)` breakage probably isn't actually a result of the mongoid upgrade itself, but because of the mongoid-enum 0.3.0 -> 0.4.0 upgrade. It no longer declares an `alias_attribute` for the field name: https://github.com/thetron/mongoid-enum/commit/2288ee6cecd02db42bd8b873d960a6d332e25bd3

**Comment by jakolehm** [comment]
Rebased.

**Comment by jakolehm** [comment]
@jnummelin @SpComb PTAL

**Comment by SpComb** [comment]
The `GridServiceDeploy#abort!` still sets `:deploy_state` (and the specs for it still seem to pass without the `reload`)

**Comment by SpComb** [comment]
The `Agent::NodeUnplugger` still logs warnings

```
I, [2017-05-12T16:08:59.900223 #20808]  INFO -- WebsocketBackend: node CAR3:26EI:CKHW:M3US:6L3D:W7CI:Y6KU:CJ5D:6ERB:DUHB:2ZAB:6Y3Q connection closed
I, [2017-05-12T16:08:59.901655 #20808]  INFO -- Agent::NodeUnplugger: disconnect node core-02
W, [2017-05-12T16:08:59.920551 #20808]  WARN -- : MONGODB | Unsupported client option 'safe'. It will be ignored.
W, [2017-05-12T16:08:59.921959 #20808]  WARN -- : MONGODB | Unsupported client option 'safe'. It will be ignored.
W, [2017-05-12T16:08:59.923343 #20808]  WARN -- : MONGODB | Unsupported client option 'safe'. It will be ignored.
W, [2017-05-12T16:08:59.926438 #20808]  WARN -- : MONGODB | Unsupported client option 'safe'. It will be ignored.
W, [2017-05-12T16:08:59.927775 #20808]  WARN -- : MONGODB | Unsupported client option 'safe'. It will be ignored.
W, [2017-05-12T16:08:59.928991 #20808]  WARN -- : MONGODB | Unsupported client option 'safe'. It will be ignored.
...
```

**Comment by jnummelin** [comment]
Let's merge this now to get it properly battle-tested. Fix the two remaining issues in separate PRs

**Comment by kke** [comment]
🍹

**Comment by jnummelin** [comment]
> Let's merge this now to get it properly battle-tested. Fix the two remaining issues in separate PRs

I was bit too eager. The agent spec failures are legit spec failures, need to fix those at least first.

**Pull request merged**